stan/math/rev/mat/fun/softmax.hpp

```diff
@@ -4,81 +4,78 @@
 #include <stan/math/prim/arr/err/check_nonzero_size.hpp>
 #include <stan/math/prim/mat/fun/Eigen.hpp>
 #include <stan/math/prim/mat/fun/softmax.hpp>
-#include <stan/math/rev/core.hpp>
+#include <stan/math/rev/mat.hpp>
 #include <vector>
 
 namespace stan {
 namespace math {
 
 namespace {
-class softmax_elt_vari : public vari {
- private:
-  vari** alpha_;
-  const double* softmax_alpha_;
-  const int size_;  // array sizes
-  const int idx_;   // in in softmax output
-
+class softmax_op {
+  int N;
+  double* y_mem_;  // Holds the results of the softmax
  public:
-  softmax_elt_vari(double val, vari** alpha, const double* softmax_alpha,
-                   int size, int idx)
-      : vari(val),
-        alpha_(alpha),
-        softmax_alpha_(softmax_alpha),
-        size_(size),
-        idx_(idx) {}
-  void chain() {
-    for (int m = 0; m < size_; ++m) {
-      if (m == idx_) {
-        alpha_[m]->adj_
-            += adj_ * softmax_alpha_[idx_] * (1 - softmax_alpha_[m]);
-      } else {
-        alpha_[m]->adj_ -= adj_ * softmax_alpha_[idx_] * softmax_alpha_[m];
-      }
+  /*
+   * Compute the softmax of the unconstrained input vector
+   *
+   * @param alpha Unconstrained input vector.
+   * @return Softmax of the input.
+   */
+  Eigen::VectorXd apply(const Eigen::VectorXd& alpha) {
+    N = alpha.size();
+    y_mem_ = ChainableStack::instance().memalloc_.alloc_array<double>(N);
+
+    auto y = softmax(alpha);
+    for (int n = 0; n < N; ++n)
+      y_mem_[n] = y(n);
+    return y;
+  }
+
+  /*
+   * Compute the result of multiply the transpose of the adjoint vector times
+   * the Jacobian of the softmax operator. It is more efficient to do this
+   * without actually computing the Jacobian and doing the vector-matrix
+   * product.
+   *
+   * @param adj Eigen::VectorXd of adjoints at the output of the softmax
+   * @return Eigen::VectorXd of adjoints propagated through softmax operation
+   */
+  Eigen::VectorXd multiply_adjoint_jacobian(const Eigen::VectorXd& adj) const {
```

Comment:
I haven't come up with a super-glamorous way of specifying the multiple argument version of this. Cause for the multiple argument version there's a multiply_adjoint_jacobian for each input argument.

Maybe something as simple as asking for definitions of multiply_adjoint_jacobian_first_argument, multiply_adjoint_jacobian_second_argument, multiply_adjoint_jacobian_third_argument is the way to go.

Reply:
I've been working on a design for this, but let's take this into a separate issue.  There's a brute force solution with tuples and then there's a clever typelist version.  My general rule is stay simple until the fifth or sixth ime you need to do something with templates---that is, a way higher threshold than with ordinary code for trying to lay down abstractions.